### PR TITLE
Integrate Android integration tests into CI pipeline

### LINKED ISSUE #176: Run integration tests for Android in our CI

### PR DESCRIPTION
**Pull request by nilsreichardt**
Closes #176

### CODE REVIEW

**Review by Jonas-Sander — COMMENTED** [review]
See above.

Maybe also add a comment that we might chache the AVD image to make it start up faster? (See emulator-runner docs)?

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -79,6 +79,50 @@ jobs:
...
+          api-level: ${{ matrix.api-level }}
+          target: ${{ matrix.target }}
+          arch: ${{ matrix.arch }}
+          profile: pixel_3a
+          emulator-options: -no-snapshot -no-window -no-boot-anim -camera-back virtualscene -camera-front emulated
+          working-directory: ./app
+          # We can not use a multiline command because the
+          # "android-emulator-runner" doesn't support it. It just takes every
+          # line as separate command.
+          script: flutter test integration_test --flavor dev --dart-define USER_1_EMAIL=$USER_1_EMAIL --dart-define USER_1_PASSWORD=$USER_1_PASSWORD
```
[the remark this answers is not in the document]
Comment (nilsreichardt) [reply .github/workflows/main.yml:141]: Yes, we can integrate this later in the CLI 👍 But I would like to do a rework of the CLI first and do things like better styling when printing to the console and writing much more tests.

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -79,6 +79,50 @@ jobs:
       
       - name: Run tests via "sz test"
         run: sz test -v
+
+  android-integration-test:
+    # Is recommended to run job with macOS to take advantage of hardware
+    # acceleration support provided by HAXM.
```
[the remark this answers is not in the document]
Comment (nilsreichardt) [reply .github/workflows/main.yml]: Will do 👍

**Comment by nilsreichardt** [comment]
Tests are failing because of:
```

══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following MissingPluginException was thrown running a test:
MissingPluginException(No implementation found for method Messaging#requestPermission on channel
plugins.flutter.io/firebase_messaging)

When the exception was thrown, this was the stack:
#0      convertPlatformException (package:firebase_messaging_platform_interface/src/method_channel/utils/exception.dart:13:5)
#1      MethodChannelFirebaseMessaging.requestPermission (package:firebase_messaging_platform_interface/src/method_channel/method_channel_messaging.dart:286:13)
<asynchronous suspension>
<asynchronous suspension>
(elided one frame from package:stack_trace)

The test description was:
  User should be able to sign in
════════════════════════════════════════════════════════════════════════════════════════════════════
```

Hmm, it seems so that we need to upgrade Firebase Messaging and Firebase Core to unblock this PR: https://stackoverflow.com/questions/69880191/missingpluginexceptionno-implementation-found-for-method-messagingrequestpermi

**Comment by nilsreichardt** [comment]
@Jonas-Sander FYI: I run the integration 30x times to check if they are flaky. Here are results:
* 3x of 30 builds: Android emulator difficulties to start correctly
* 1x of 30 builds: Android build needed longer than 12 minutes to build (same problem as we have with iOS)

I think in these cases, we should just restart the CI manually.

image

https://github.com/SharezoneApp/sharezone-app/runs/6881588330?check_suite_focus=true

File: app/lib/blocs/homework/homework_page_bloc.dart [hunk 3]
```diff
@@ -128,7 +128,14 @@ class HomeworkPageBloc extends BlocBase {
...
   /// Take care of closing streams.
   @override
-  void dispose() {
+  Future<void> dispose() async {
+    await Future.wait([
+      _homeworkNotDoneSubject.drain(),
+      _homeworkDoneSubject.drain(),
+      _errorsSubject.drain(),
+      _homeworkListSubject.drain(),
+    ]);
```
Comment (nilsreichardt) [inline app/lib/blocs/homework/homework_page_bloc.dart:137]: Needed to `drain` the streams. Otherwise, would be the CI ~50% flaky: https://github.com/SharezoneApp/sharezone-app/runs/6879559070?check_suite_focus=true

```
══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
The following StateError was thrown while finalizing the widget tree:
Bad state: You cannot close the subject while items are being added from addStream
When the exception was thrown, this was the stack:
#0      Subject.close (package:rxdart/src/subjects/subject.dart:153:7)
#1      HomeworkPageBloc.dispose (package:sharezone/blocs/homework/homework_page_bloc.dart:137:26)
#2      _BlocProviderState.dispose (package:bloc_provider/src/bloc_provider.dart:53:17)
#3      StatefulElement.unmount (package:flutter/src/widgets/framework.dart:4983:11)
#4      _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1926:13)
#5      _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1924:7)
#6      ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:4807:14)
#7      _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1922:13)
#8      _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:1924:7)
#9      ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:4807:14)
#10     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:1922:13)
```

File: app/lib/notifications/firebase_messaging_callback_configurator.dart [hunk 4]
```diff
@@ -126,6 +127,16 @@ class FirebaseMessagingCallbackConfigurator {
...
 Future<void> _requestIOSPermission(BuildContext context) async {
+  const isIntegrationTest = bool.fromEnvironment('IS_INTEGRATION_TEST');
+  final isAndroid = PlatformCheck.isAndroid;
+  if (isIntegrationTest && isAndroid) {
+    // We skip requesting the permission in the integration test for Android
+    // because when running the tests with GitHub Actions Android emulators, the
+    // following exception will be thrown: "MissingPluginException(No
+    // implementation found for method Messaging#requestPermission on channel".
+    return;
+  }
```
Comment (nilsreichardt) [inline app/lib/notifications/firebase_messaging_callback_configurator.dart:138]: Here the build logs: https://github.com/SharezoneApp/sharezone-app/runs/6868516775?check_suite_focus=true

File: .github/workflows/main.yml [hunk 5]
```diff
@@ -14,6 +14,10 @@ name: CI
 on:
   # Triggers the workflow on pull request events
   pull_request:
+    # The types trigger are needed to run the jobs again which are skipped for
+    # draft pull requests.
+    # https://github.community/t/dont-run-actions-on-draft-pull-requests/16817/16
+    types: [opened, synchronize, reopened, ready_for_review]
```
Comment (Jonas-Sander) [inline .github/workflows/main.yml:20]: So either we only use this and remove `if: ${{ github.event.pull_request.draft == false }}` or the other way around.

For right now I would keep `if: ${{ github.event.pull_request.draft == false }}` and if we decide to change this then do it in a separate PR. 

Anyways we do have actions right now that always run and for which it makes sense for me like `check-files-licence-headers`.

File: .github/workflows/main.yml [hunk 6]
```diff
@@ -112,7 +116,57 @@ jobs:
...
+    # https://github.com/ReactiveCircus/android-emulator-runner#usage
+    runs-on: macos-latest
+    if: ${{ github.event.pull_request.draft == false }}
+    timeout-minutes: 30
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-java@v3
+        with:
+          distribution: 'temurin'
+          java-version: '11'
```
Comment (Jonas-Sander) [inline .github/workflows/main.yml:134]: For what is this exactly needed again? Couldn't find anything mentioning this step at https://github.com/ReactiveCircus/android-emulator-runner

File: .github/workflows/main.yml [hunk 7]
```diff
@@ -112,7 +116,57 @@ jobs:
...
+          flutter-version: ${{ env.FLUTTER_VERSION }}
+          channel: ${{ env.FLUTTER_CHANNEL }}
+      
+      - name: Run integration tests
+        uses: reactivecircus/android-emulator-runner@v2
+        env:
+          USER_1_EMAIL: ${{ secrets.INTEGRATION_TEST_USER_1_EMAIL }}
+          USER_1_PASSWORD: ${{ secrets.INTEGRATION_TEST_USER_1_PASSWORD }}
+        # We are using the "google_atd" target because it's designed for
+        # automated testing.
```
Comment (Jonas-Sander) [inline .github/workflows/main.yml:150]:
```suggestion
        # We are using the "google_atd" ("Automated Test Device") target because it's designed for
        # automated testing and thus uses fewer resources.
```

File: .github/workflows/main.yml [hunk 8]
```diff
@@ -112,7 +116,57 @@ jobs:
...
+        #
+        # See
+        # https://android-developers.googleblog.com/2021/10/whats-new-in-scalable-automated-testing.html.
```
Comment (Jonas-Sander) [inline .github/workflows/main.yml:153]:
```suggestion
        # See "Slimmer Emulator System Images" / "Automated Test Device" at:
        # https://android-developers.googleblog.com/2021/10/whats-new-in-scalable-automated-testing.html.
```

File: .github/workflows/main.yml [hunk 9]
```diff
@@ -112,7 +116,57 @@ jobs:
...
+        # Because of the use of "google_atd" we are bit limited in the choice of
+        # API level, arch and channel. See here all requirements:
+        # https://github.com/ReactiveCircus/android-emulator-runner#configurations 
+        with:
+          api-level: 30
+          target: google_atd
+          arch: x86
+          profile: pixel_3a
+          channel: canary
+          emulator-options: -no-snapshot -no-window -no-boot-anim -camera-back virtualscene -camera-front emulated
```
Comment (Jonas-Sander) [inline .github/workflows/main.yml:164]: It might be helpful to know why these emulator-options are used. You could also put some together and explain only their general purpose since one can probably google the exact description.
```
-no-snapshot -no-window -no-boot-anim -> for faster startup and execution (I guess?)
-camera-back virtualscene -> TODO
-camera-front emulated -> TODO
```

**Review by Jonas-Sander — CHANGES_REQUESTED** [review]

File: app/lib/notifications/firebase_messaging_callback_configurator.dart (hunk 4 above)
Comment (nilsreichardt) [reply app/lib/notifications/firebase_messaging_callback_configurator.dart:138]: We already have the permission because there is no need to request the permission on Android devices. This is also the reason why the method is called `_requestIOSPermission()`. But we need to rename or adjust this method for Android 13. Android 13 will introduce a similar feature as iOS has: you need to request the permission for sending push notification. 

https://techcrunch.com/2022/03/17/google-launches-android-13-developer-preview-2-psuh-notifications-opt-in/

I think the best solution would be to check if we need to call the `requestPermission` method on this device (based on the OS version). This would also remove the current solution with `isIntegrationTest`. But I would do this later when we adjust our app to Android 13.

File: .github/workflows/main.yml (hunk 5 above)
Comment (nilsreichardt) [reply .github/workflows/main.yml:20]: My comment is a bit misleading. We need both. `if: ${{ github.event.pull_request.draft == false }}` and the types. The reason for this is:

If we have a draft pull request, all jobs with `if: ${{ github.event.pull_request.draft == false }}` will be skipped. When you switch from draft pull request to normal pull request, the jobs will **not** re-trigger (until you push something again). Therefore, it would be possible to have a draft pull request where the jobs will be skipped, switch to normal pull request and then merge into `main` without running our CI.

`types: [opened, synchronize, reopened, ready_for_review]` will re-trigger the jobs when you switch from draft pull request to normal pull request.

I will adjust the comment later 👍

File: .github/workflows/main.yml [hunk 10]
```diff
@@ -112,7 +116,57 @@ jobs:
...
+          USER_1_PASSWORD: ${{ secrets.INTEGRATION_TEST_USER_1_PASSWORD }}
+        # We are using the "google_atd" target because it's designed for
+        # automated testing.
+        #
+        # See
+        # https://android-developers.googleblog.com/2021/10/whats-new-in-scalable-automated-testing.html.
+        #
+        # Because of the use of "google_atd" we are bit limited in the choice of
+        # API level, arch and channel. See here all requirements:
+        # https://github.com/ReactiveCircus/android-emulator-runner#configurations
```
[the remark this answers is not in the document]
Comment (nilsreichardt) [reply .github/workflows/main.yml:157]: Interesting question. Can't test this right now, because integration screenshot testing isn't working yet, see https://github.com/flutter/flutter/issues/103222

File: app/lib/notifications/firebase_messaging_callback_configurator.dart (hunk 4 above)
Comment (Jonas-Sander) [reply app/lib/notifications/firebase_messaging_callback_configurator.dart:138]: Wait but why do you even call isIntegrationTest in the method at all? And why don't we just do `if(platform== iOS)` _requestIOSPermissions? Or replace `kIsDriverTest` further above in the file with `isIntegrationTest`?

File: .github/workflows/main.yml (hunk 9 above)
Comment (nilsreichardt) [reply .github/workflows/main.yml:165]: I will add comments later 👍

File: app/lib/blocs/homework/homework_page_bloc.dart (hunk 3 above)
Comment (nilsreichardt) [reply app/lib/blocs/homework/homework_page_bloc.dart:137]: Let's see if this makes problems. If yes, we need to find a different solution.

File: .github/workflows/main.yml (hunk 6 above)
Comment (Jonas-Sander) [reply .github/workflows/main.yml:134]: Can you add this via a comment above it?

File: app/lib/notifications/firebase_messaging_callback_configurator.dart (hunk 4 above)
Comment (nilsreichardt) [reply app/lib/notifications/firebase_messaging_callback_configurator.dart:138]: I would suggest doing an `if(PlatformCheck.isIOS)` in the `_requestIOSPermissions()`

File: .github/workflows/main.yml (hunk 6 above)
Comment (nilsreichardt) [reply .github/workflows/main.yml:137]: done

File: .github/workflows/main.yml (hunk 5 above)
Comment (nilsreichardt) [reply .github/workflows/main.yml]: Done ✅

File: .github/workflows/main.yml [hunk 11]
```diff
@@ -112,7 +122,78 @@ jobs:
...
+          #   the back camera where you can walk. We are using this option
+          #   because most emulator have this option enabled when you create you
+          #   emulator via Android Studio. Can be helpful for testing things
+          #   with the camera. It's also possible to inject images into the
+          #   virtual scene to test things like qr code scanner.
+          # * "-camera-front": Shows this green monster when opening the front
+          #   camera. We are using this option because most emulator have this
+          #   option enabled when you create your emulator via Android Studio.
+          #   Can be helpful to test things with the camera.
+          emulator-options: -no-snapshot -no-window -no-boot-anim -camera-back virtualscene -camera-front emulated
```
Comment (Jonas-Sander) [inline .github/workflows/main.yml:191]: Would it also work without these? Do have to specify them? If we have no use for them right now I would leave them out.

File: .github/workflows/main.yml [hunk 12]
```diff
@@ -154,7 +235,7 @@ jobs:
             USER_1_EMAIL=$USER_1_EMAIL \
             --dart-define \
             USER_1_PASSWORD=$USER_1_PASSWORD \
-            -d $SIMULATOR_UDID
+            --dart-define
```
Comment (Jonas-Sander) [inline .github/workflows/main.yml:239]: Why the empty `--dart-define`?

Comment (Jonas-Sander) [reply .github/workflows/main.yml:239]: And why do you change this (inside iOS action) in this PR?

File: .github/workflows/main.yml (hunk 11 above)
Comment (nilsreichardt) [reply .github/workflows/main.yml:191]: It should also work without the camera parameter 👍 

But we need to remember to set these again when we need these (or assume this person has this knowledge) because as I mentioned has the Android Emulators created with Android Studio by default these options. When I implemented the integration tests for the qr code scanner took me this a few hours to figure out why the integration tests are working locally but no with GitHub Actions.

I will remove the options 👍

File: .github/workflows/main.yml (hunk 12 above)
Comment (nilsreichardt) [reply .github/workflows/main.yml]: was a mistake

File: .github/workflows/main.yml (hunk 11 above)
Comment (Jonas-Sander) [reply .github/workflows/main.yml]: Ah I misunderstood your comments then. Please re-add them (sorry!) but change them a bit.   

Instead of 
> Can be helpful for testing things with the camera.

Write : Is **needed** to test things with the camera

And instead of 
> We are using this option because most emulator have this option enabled when you create you emulator via Android Studio.

Write something along the lines of: Default option for emulators created by Android Studio. We use this so we have the same environment locally as when using this action.

Does this make sense?

**Comment by nilsreichardt** [comment]
Making this as a draft again to check if the "Timeout waiting for emulator to boot." flaky error can be fixed.

**Review by Jonas-Sander — APPROVED** [review]
LGTM

**Comment by Jonas-Sander** [comment]
Just want to say that some android actions fail but because of an firebase auth error, e.g. https://github.com/SharezoneApp/sharezone-app/actions/runs/3160777863/jobs/5145617181

**Comment by nilsreichardt** [comment]
Using `flutter test` instead of `flutter drive` seams to fix the problem with the Android Emulator. Instead of 30% - 40% failure rate we now have only ~6%. I'm testing the `google_atd`. When there is everything fine, we can merge this :)

**Comment by Sharezone-Admin** [comment]
⬇️ Generated builds by [Codemagic](https://codemagic.io/app/629c82ea463af7ff553fc7a5/build/63f7ef97304e8dfd58447b7c) for commit 817f919a4d2744bce34e51bce70d718aa49eeb54 ⬇️

_Note_: Only Sharezone team members are able to install the iOS app.

| iOS | Android |
|:-:|:-:|
| image <br /> [Download link](https://api.codemagic.io/artifacts/3647a207-714a-46e6-bf65-31b436ffa8f7/c6cf675d-4b0c-4f69-a761-2fecbf57ee07/sharezone.ipa) | image <br /> [Download link](https://api.codemagic.io/artifacts/5ba1cfa4-f6c7-4f52-9091-0619ebfa82cb/f6ed5dd1-24e6-4009-bf7b-1ec87cde2427/app-dev-release.apk) |

File: .github/workflows/main.yml [hunk 13]
```diff
@@ -131,7 +131,104 @@ jobs:
...
+          # We can not use a multiline command because the
+          # "android-emulator-runner" action doesn't support it. It just takes every
+          # line as separate command.
+          #
+          # We use the `flutter drive` instead of the `flutter test` command
+          # because the test command times out after 12 minutes. But building the
+          # app takes more than 12 minutes... It seems so that there is no way to
+          # set the timeout (the --timeout argument has not effect). Tracking
+          # issue: https://github.com/flutter/flutter/issues/105913
+          script: flutter test integration_test --flavor dev --dart-define USER_1_EMAIL=$USER_1_EMAIL --dart-define USER_1_PASSWORD=$USER_1_PASSWORD
```
Comment (Jonas-Sander) [inline .github/workflows/main.yml:231]: The comment is out-of-date (you're using `flutter test` here, not `flutter drive`)

File: .github/workflows/main.yml [hunk 14]
```diff
@@ -131,7 +131,104 @@ jobs:
...
+          #   the back camera where you can walk. Default option for emulators
+          #   created by Android Studio. We use this so we have the same
+          #   environment locally as when using this action. Is needed when
+          #   testing things with the camera. It's also possible to inject
+          #   images into the virtual scene to test things like qr code scanner.
+          # * "-camera-front": Shows this green monster when opening the front
+          #   camera. Default option for emulators created by Android Studio. We
+          #   use this so we have the same environment locally as when using
+          #   this action. Is needed when testing things with camera.
+          emulator-options: -no-snapshot -no-window -no-boot-anim -camera-back virtualscene -camera-front emulated
```
Comment (Jonas-Sander) [inline .github/workflows/main.yml:220]: Teeny tiny nit: Could you use the same order for the `emulator-options:` flags and for the comment above explaining them?

**Review by Jonas-Sander — APPROVED** [review]
See nits above.

File: .github/workflows/main.yml (hunk 14 above)
Comment (nilsreichardt) [reply .github/workflows/main.yml]: Changed

File: .github/workflows/main.yml (hunk 13 above)
Comment (nilsreichardt) [reply .github/workflows/main.yml]: ah, thanks
